### DISCUSSION TITLE
99. 데드락, 대기 트랜잭션을 InnoDB에서 확인하는 방법은 무엇인가요?

### DISCUSSION THREAD

**Discussion by jongmee** [jewelry-box/cs-interview#103] [General]
데드락, 대기 트랜잭션을 InnoDB에서 확인하는 방법은 무엇인가요?

**Comment by linirini** [comment]
### 1. 데드락 확인
InnoDB는 데드락이 발생하면 가장 최근에 발생한 데드락 정보를 저장한다.
#### 방법 1: SHOW ENGINE INNODB STATUS
MySQL 명령어로 데드락 정보를 확인할 수 있다.
``` sql
SHOW ENGINE INNODB STATUS;
```
출력 내용 중 LATEST DETECTED DEADLOCK 섹션에 데드락 관련 정보가 표시된다.
어떤 트랜잭션이 충돌했는지, 어떤 쿼리가 원인이었는지 상세히 나온다.
### 방법 2: MySQL Performance Schema 활용
Performance Schema에서 데드락 정보를 조회할 수도 있다.
``` sql
SELECT * FROM performance_schema.events_transactions_current WHERE STATE = 'DEADLOCK';
```
### 2. 대기 중인 트랜잭션 확인
#### 방법 1: SHOW PROCESSLIST
현재 실행 중인 프로세스를 확인하여 대기 중인 트랜잭션을 찾을 수 있습니다.
``` sql
SHOW FULL PROCESSLIST;
```
Command가 Sleep 또는 Locked 상태인 프로세스가 트랜잭션 대기 중일 가능성이 높다. Time 필드가 높으면 오래된 대기 트랜잭션일 수 있다.
#### 방법 2: INFORMATION_SCHEMA 활용
INNODB_LOCKS와 INNODB_LOCK_WAITS 테이블에서 대기 중인 락 정보를 조회할 수 있다.
``` sql
SELECT * FROM INFORMATION_SCHEMA.INNODB_LOCKS;
SELECT * FROM INFORMATION_SCHEMA.INNODB_LOCK_WAITS;
```
INNODB_LOCKS는 현재 잡혀 있는 락 정보를 보여준다.
INNODB_LOCK_WAITS는 어떤 트랜잭션이 어떤 락을 대기 중인지 보여준다.
#### 방법 3: Performance Schema 활용
``` sql
SELECT * FROM performance_schema.data_locks;
SELECT * FROM performance_schema.threads WHERE PROCESSLIST_STATE = 'Waiting for table metadata lock';
```
data_locks 테이블은 락 정보를 제공하며, threads 테이블의 PROCESSLIST_STATE 컬럼으로 대기 상태 확인할 수 있다.
